File: 2015/go/not-quite-lisp/main.go
// https://adventofcode.com/2015/day/1

package main

import "io/ioutil"

func readInput() (string, error) {
	data, err := ioutil.ReadFile("input.txt")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func part1(input string) int {
	floor := 0
	for _, c := range input {
		switch c {
		case '(':
			floor += 1
		case ')':
			floor -= 1
		}
	}
	return floor
}

func part2(input string) int {
	floor := 1
	for i, c := range input {
		switch c {
		case '(':
			floor += 1
		case ')':
			floor -= 1
		}
		if floor < 0 {
			return i
		}
	}
	return -1
}

func main() {
	input, err := readInput()
	if err != nil {
		panic(err)
	}

	println("Part 1", part1(input))
	println("Part 2", part2(input))
}

File: 2015/rs/not-quite-lisp/src/main.rs
use std::fs;

fn read_input() -> Result<String, std::io::Error> {
    fs::read_to_string("input.txt")
}

fn char2increment(c: char) -> i32 {
    match c {
        ')' => -1,
        '(' => 1,
        _ => panic!("Unexpected character in input: {}", c),
    }
}

fn part1(input: &String) -> i32 {
    let mut floor = 0;
    for c in input.chars() {
        floor += char2increment(c);
    }
    floor
}

fn part2(input: &String) -> usize {
    let mut floor = 0;
    for (i, c) in input.chars().enumerate() {
        floor += char2increment(c);
        if floor == -1 {
            return i;
        }
    }
    0
}

fn main() {
    let input = read_input().unwrap();
    println!("Part one: {}", part1(&input));
    println!("Part two: {}", part2(&input));
}
